src/main/java/io/github/nucleuspowered/nucleus/modules/world/listeners/WorldListener.java

```diff
@@ -4,39 +4,57 @@
  */
 package io.github.nucleuspowered.nucleus.modules.world.listeners;
 
-import io.github.nucleuspowered.nucleus.Nucleus;
 import io.github.nucleuspowered.nucleus.internal.ListenerBase;
 import io.github.nucleuspowered.nucleus.internal.PermissionRegistry;
-import io.github.nucleuspowered.nucleus.internal.annotations.ConditionalListener;
 import io.github.nucleuspowered.nucleus.modules.world.WorldModule;
 import io.github.nucleuspowered.nucleus.modules.world.config.WorldConfig;
 import io.github.nucleuspowered.nucleus.modules.world.config.WorldConfigAdapter;
+import org.spongepowered.api.Sponge;
 import org.spongepowered.api.entity.living.player.Player;
 import org.spongepowered.api.event.Listener;
 import org.spongepowered.api.event.entity.MoveEntityEvent;
 import org.spongepowered.api.event.filter.Getter;
+import org.spongepowered.api.scheduler.Task;
+import org.spongepowered.api.world.Location;
 import org.spongepowered.api.world.World;
+import org.spongepowered.api.world.teleport.TeleportHelperFilters;
 
-import java.util.function.Predicate;
+import java.util.Optional;
+import java.util.function.Consumer;
 
-@ConditionalListener(WorldListener.Condition.class)
-public class WorldListener extends ListenerBase {
+public class WorldListener extends ListenerBase implements ListenerBase.Conditional {
 
-    @Listener
-    public void onPlayerTeleport(MoveEntityEvent.Teleport event, @Getter("getTargetEntity") Player player) {
-        World target = event.getToTransform().getExtent();
-        if (player.getWorld().equals(target)) return;
+	@Listener
+	public void onPlayerTeleport(MoveEntityEvent.Teleport event, @Getter("getTargetEntity") Player player) {
+		World target = event.getToTransform().getExtent();
+		if (player.getWorld().equals(target)) return;
 
-        if (!player.hasPermission(PermissionRegistry.PERMISSIONS_PREFIX + "worlds." + target.getName().toLowerCase())) {
-            event.setCancelled(true);
-            player.sendMessage(plugin.getMessageProvider().getTextMessageWithFormat("world.access.denied", target.getName()));
-        }
-    }
+		if (!player.hasPermission(PermissionRegistry.PERMISSIONS_PREFIX + "worlds." + target.getName().toLowerCase())) {
+			event.setCancelled(true);
+			player.sendMessage(plugin.getMessageProvider().getTextMessageWithFormat("world.access.denied", target.getName()));
+			if (event instanceof MoveEntityEvent.Teleport.Portal) {
+				Sponge.getScheduler().createTaskBuilder()
+					.delayTicks(1)
+					.execute(relocate(player))
+					.submit(plugin);
+			}
+		}
+	}
 
-    public static class Condition implements Predicate<Nucleus> {
+	@Override
+	public boolean shouldEnable() {
+		return plugin.getConfigValue(WorldModule.ID, WorldConfigAdapter.class, WorldConfig::isSeparatePermissions).orElse(false);
+	}
+
+	private Consumer<Task> relocate(Player player) {
+		return src -> {
```

Comment:
Would it be better to store this as a global static variable, and then in `execute(Consumer)` run `this.relocate.apply(player)`? Just thinking about object creation, that's all!

Reply:
I've only ever done `Consumer<Task>`s as a method or inline with a lambda. I wasn't aware there were other alternatives.  When it comes to performance I know you have a better grasp then I do so I'll follow your lead.

Reply:
Yeah, if you think of this consumer as an object that you can reuse (it's effectively a one method class you're constructing each time), then we just store it.

You can do `private final static Consumer<Task> relocate = src -> { magic(); };`, and then call `this.relocate.apply(src)`.

Reply:
Not really your fault, I named it improperly.